### PR TITLE
Rib 161 implement user profile page

### PR DESCRIPTION
**Pull request by vdrfg**
### RIB-161: Implementation of User's Profile

- User can access his profile from top bar menu.
- Endpoint takes username for now since they are unique and ignore cases and other users can easily "search" for a user.
- Default message for user bio.
- User can add a bio/description by clicking into the field. Default message disappears. Text is saved when clicked away. 
- Follow button follows user and dynamically changes the follower count.
- Follow button is not visible for anonymous users and for users on their own profile.
- Page loads all interests with ratings from user, sorted by number of total likes.
- Interest can be extended by click and shows up to 3 best rated places, "Details" button will lead to separate page that lists all places.
- Page dynamically changes according to screen size.
- Small screens have slightly different positioning of Follow and Details buttons.

image
image
image
image
image

### CODE REVIEW

File: src/main/java/it/rate/webapp/services/RatingService.java [hunk 1]
```diff
@@ -88,4 +93,52 @@ private Criterion getCriterion(Long criterionId) {
...
+              return new UserRatedInterestDTO(
+                  interest.getId(),
+                  interest.getName(),
+                  interest.getImageName(),
+                  interest.countLikes(),
+                  userRatedPlaceDTOs);
+            })
+        .sorted(Comparator.comparingInt(UserRatedInterestDTO::likes).reversed())
+        .toList();
+  }
```
Comment (marthubner) [inline src/main/java/it/rate/webapp/services/RatingService.java:143]: This should probably be distributed to multiple smaller methods and possibly even services but we can do it later.

File: src/main/java/it/rate/webapp/services/RatingService.java [hunk 2]
```diff
@@ -88,4 +93,52 @@ private Criterion getCriterion(Long criterionId) {
...
+                          })
+                      .sorted(Comparator.comparingDouble(UserRatedPlaceDTO::avgRating).reversed())
+                      .collect(Collectors.toList());
+
+              return new UserRatedInterestDTO(
+                  interest.getId(),
+                  interest.getName(),
+                  interest.getImageName(),
+                  interest.countLikes(),
+                  userRatedPlaceDTOs);
```
Comment (marthubner) [inline src/main/java/it/rate/webapp/services/RatingService.java:139]: There is a constructor for this that takes in an interest and is not used

File: src/main/java/it/rate/webapp/services/RatingService.java [hunk 3]
```diff
@@ -88,4 +93,52 @@ private Criterion getCriterion(Long criterionId) {
...
+                                            new UserRatingDTO(
+                                                userRating.getRating(),
+                                                userRating.getCriterion().getName()))
+                                    .collect(Collectors.toList());
+
+                            return new UserRatedPlaceDTO(
+                                place.getId(),
+                                place.getName(),
+                                place.getAverageRating() / 2,
+                                userRatingDTOs);
```
Comment (marthubner) [inline src/main/java/it/rate/webapp/services/RatingService.java:129]: Constructor which takes place can be made to make this a bit shorter

File: src/main/resources/static/scripts/userBioEdit.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
+document.addEventListener('DOMContentLoaded', function () {
+    const userBio = document.getElementById('user-bio');
+    const maxBioLength = 150;
```
Comment (marthubner) [inline src/main/resources/static/scripts/userBioEdit.js:3]: this should not be hardcoded

**Comment by marthubner** [comment]
image
There's an error in javascript when user is viewing his own profile

File: src/main/java/it/rate/webapp/services/RatingService.java [hunk 5]
```diff
@@ -88,4 +93,30 @@ private Criterion getCriterion(Long criterionId) {
...
+    Map<Interest, Map<Place, List<Rating>>> ratingsByInterestAndPlace =
+        userRatings.stream()
+            .collect(
+                Collectors.groupingBy(
+                    rating -> rating.getPlace().getInterest(),
+                    Collectors.groupingBy(Rating::getPlace, Collectors.toList())));
+
+    return ratingsByInterestAndPlace.entrySet().stream()
+        .map(
+            interest ->
```
Comment (MatejFrnka) [inline src/main/java/it/rate/webapp/services/RatingService.java:109]: refactor

**Review by MatejFrnka — CHANGES_REQUESTED** [review]

**Review by MatejFrnka — APPROVED** [review]

**Review by marthubner — APPROVED** [review]

**Review by BenediktVitek — APPROVED** [review]

**Review by dondowoldo — CHANGES_REQUESTED** [review]
Looks good and not sure if this is supposed to be a part of this pull request, but with the user profile implemented, i would like the user instances to be clickable across our website and redirecting to the users page. I think this would be for "Manage Users" and "Applications" in Interest Admin panel.

Also user for example can currently create an account with a username "signup" and nobody will ever see his profile 🤣 Maybe we need to have a look at that as well.

**Pull request merged**